src/@next/pages/CheckoutPage/subpages/CheckoutAddressSubpage.tsx

```diff
@@ -23,58 +23,84 @@ export interface ICheckoutAddressSubpageHandles {
 
 interface IProps extends RouteComponentProps<any> {
   changeSubmitProgress: (submitInProgress: boolean) => void;
+  onSubmitSuccess: () => void;
 }
 
 const CheckoutAddressSubpageWithRef: RefForwardingComponent<
   ICheckoutAddressSubpageHandles,
   IProps
-> = ({ changeSubmitProgress, ...props }: IProps, ref) => {
-  const checkoutAddressFormId = "address-form";
-  const checkoutAddressFormRef = useRef<HTMLFormElement>(null);
+> = ({ changeSubmitProgress, onSubmitSuccess, ...props }: IProps, ref) => {
+  const checkoutShippingAddressFormId = "shipping-address-form";
+  const checkoutShippingAddressFormRef = useRef<HTMLFormElement>(null);
+  const checkoutBillingAddressFormId = "billing-address-form";
+  const checkoutBillingAddressFormRef = useRef<HTMLFormElement>(null);
   const checkoutNewAddressFormId = "new-address-form";
 
-  useImperativeHandle(ref, () => ({
-    submitAddress: () => {
-      if (user && selectedShippingAddressId) {
-        checkoutAddressFormRef.current?.dispatchEvent(
-          new Event("submit", { cancelable: true })
-        );
-      } else {
-        // TODO validate form
-        checkoutAddressFormRef.current?.dispatchEvent(
-          new Event("submit", { cancelable: true })
-        );
-      }
-    },
-  }));
-
-  const history = useHistory();
   const { user } = useAuth();
   const {
     checkout,
-    setShippingAddress,
     selectedShippingAddressId,
+    billingAsShipping,
+    setShippingAddress,
+    setBillingAddress,
+    setBillingAsShippingAddress,
   } = useCheckout();
+  const { items } = useCart();
   const { countries } = useContext(ShopContext);
 
-  const [errors, setErrors] = useState<IFormError[]>([]);
+  const [shippingErrors, setShippingErrors] = useState<IFormError[]>([]);
+  const [billingErrors, setBillingErrors] = useState<IFormError[]>([]);
 
   const intl = useIntl();
 
+  const isShippingRequiredForProducts =
+    items &&
+    items.some(
+      ({ variant }) => variant.product?.productType.isShippingRequired
+    );
   const checkoutShippingAddress = checkout?.shippingAddress
     ? {
         ...checkout?.shippingAddress,
+        email: checkout?.email || user?.email || undefined,
```

Comment:
Does || undefined changes anything?

Reply:
Could, as `IAddress` interface didn't accept null values. But it shouldn't change anything, so I have fixed it.